File: packages/google_mlkit_barcode_scanning/android/src/main/java/com/google_mlkit_barcode_scanning/BarcodeScanner.java
package com.google_mlkit_barcode_scanning;

import android.content.Context;
import android.graphics.Point;
import android.graphics.Rect;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.mlkit.vision.barcode.common.Barcode;
import com.google.mlkit.vision.barcode.BarcodeScannerOptions;
import com.google.mlkit.vision.barcode.BarcodeScanning;
import com.google.mlkit.vision.common.InputImage;
import com.google_mlkit_commons.InputImageConverter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class BarcodeScanner implements MethodChannel.MethodCallHandler {
    private static final String START = "vision#startBarcodeScanner";
    private static final String CLOSE = "vision#closeBarcodeScanner";

    private final Context context;
    private final Map<String, com.google.mlkit.vision.barcode.BarcodeScanner> instances = new HashMap<>();

    public BarcodeScanner(Context context) {
        this.context = context;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        switch (method) {
            case START:
                handleDetection(call, result);
                break;
            case CLOSE:
                closeDetector(call);
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private com.google.mlkit.vision.barcode.BarcodeScanner initialize(MethodCall call) {
        List<Integer> formatList = call.argument("formats");
        BarcodeScannerOptions barcodeScannerOptions;
        if (formatList.size() > 1) {
            int[] array = new int[formatList.size()];
            for (int i = 1; i < formatList.size(); i++) {
                array[i] = formatList.get(i);
            }
            barcodeScannerOptions = new BarcodeScannerOptions.Builder().setBarcodeFormats(formatList.get(0), array).build();
        } else {
            barcodeScannerOptions = new BarcodeScannerOptions.Builder().setBarcodeFormats(formatList.get(0)).build();
        }
        return BarcodeScanning.getClient(barcodeScannerOptions);
    }

    private void handleDetection(MethodCall call, final MethodChannel.Result result) {
        Map<String, Object> imageData = call.argument("imageData");
        InputImage inputImage = InputImageConverter.getInputImageFromData(imageData, context, result);
        if (inputImage == null) return;

        String id = call.argument("id");
        com.google.mlkit.vision.barcode.BarcodeScanner barcodeScanner = instances.get(id);
        if (barcodeScanner == null) {
            barcodeScanner = initialize(call);
            instances.put(id, barcodeScanner);
        }

        barcodeScanner.process(inputImage)
                .addOnSuccessListener(barcodes -> {
                    List<Map<String, Object>> barcodeList = new ArrayList<>(barcodes.size());
                    for (Barcode barcode : barcodes) {

                        Map<String, Object> barcodeMap = new HashMap<>();
                        int valueType = barcode.getValueType();
                        barcodeMap.put("type", valueType);
                        barcodeMap.put("format", barcode.getFormat());
                        barcodeMap.put("rawValue", barcode.getRawValue());
                        barcodeMap.put("rawBytes", barcode.getRawBytes());
                        barcodeMap.put("displayValue", barcode.getDisplayValue());
                        barcodeMap.put("rect", getBoundingPoints(barcode.getBoundingBox()));
                        Point[] cornerPoints = barcode.getCornerPoints();
                        List<Map<String, Integer>> points = new ArrayList<>();
                        addPoints(cornerPoints, points);
                        barcodeMap.put("points", points);
                        switch (valueType) {
                            case Barcode.TYPE_UNKNOWN:
                            case Barcode.TYPE_ISBN:
                            case Barcode.TYPE_PRODUCT:
                            case Barcode.TYPE_TEXT:
                                break;
                            case Barcode.TYPE_WIFI:
                                barcodeMap.put("ssid", barcode.getWifi().getSsid());
                                barcodeMap.put("password", barcode.getWifi().getPassword());
                                barcodeMap.put("encryption", barcode.getWifi().getEncryptionType());
                                break;
                            case Barcode.TYPE_URL:
                                barcodeMap.put("title", barcode.getUrl().getTitle());
                                barcodeMap.put("url", barcode.getUrl().getUrl());
                                break;
                            case Barcode.TYPE_EMAIL:
                                barcodeMap.put("address", barcode.getEmail().getAddress());
                                barcodeMap.put("body", barcode.getEmail().getBody());
                                barcodeMap.put("subject", barcode.getEmail().getSubject());
                                barcodeMap.put("emailType", barcode.getEmail().getType());
                                break;
                            case Barcode.TYPE_PHONE:
                                barcodeMap.put("number", barcode.getPhone().getNumber());
                                barcodeMap.put("phoneType", barcode.getPhone().getType());
                                break;
                            case Barcode.TYPE_SMS:
                                barcodeMap.put("message", barcode.getSms().getMessage());
                                barcodeMap.put("number", barcode.getSms().getPhoneNumber());
                                break;
                            case Barcode.TYPE_GEO:
                                barcodeMap.put("latitude", barcode.getGeoPoint().getLat());
                                barcodeMap.put("longitude", barcode.getGeoPoint().getLng());
                                break;
                            case Barcode.TYPE_DRIVER_LICENSE:
                                barcodeMap.put("addressCity", barcode.getDriverLicense().getAddressCity());
                                barcodeMap.put("addressState", barcode.getDriverLicense().getAddressState());
                                barcodeMap.put("addressZip", barcode.getDriverLicense().getAddressZip());
                                barcodeMap.put("addressStreet", barcode.getDriverLicense().getAddressStreet());
                                barcodeMap.put("issueDate", barcode.getDriverLicense().getIssueDate());
                                barcodeMap.put("birthDate", barcode.getDriverLicense().getBirthDate());
                                barcodeMap.put("expiryDate", barcode.getDriverLicense().getExpiryDate());
                                barcodeMap.put("gender", barcode.getDriverLicense().getGender());
                                barcodeMap.put("licenseNumber", barcode.getDriverLicense().getLicenseNumber());
                                barcodeMap.put("firstName", barcode.getDriverLicense().getFirstName());
                                barcodeMap.put("lastName", barcode.getDriverLicense().getLastName());
                                barcodeMap.put("country", barcode.getDriverLicense().getIssuingCountry());
                                break;
                            case Barcode.TYPE_CONTACT_INFO:
                                barcodeMap.put("firstName", barcode.getContactInfo().getName().getFirst());
                                barcodeMap.put("lastName", barcode.getContactInfo().getName().getLast());
                                barcodeMap.put("formattedName", barcode.getContactInfo().getName().getFormattedName());
                                barcodeMap.put("organization", barcode.getContactInfo().getOrganization());
                                List<Map<String, Object>> queries = new ArrayList<>();
                                for (Barcode.Address address : barcode.getContactInfo().getAddresses()) {
                                    Map<String, Object> addressMap = new HashMap<>();
                                    addressMap.put("addressType", address.getType());
                                    List<String> addressLines = new ArrayList<>();
                                    Collections.addAll(addressLines, address.getAddressLines());
                                    addressMap.put("addressLines", addressLines);
                                    queries.add(addressMap);
                                }
                                barcodeMap.put("addresses", queries);
                                queries = new ArrayList<>();
                                for (Barcode.Phone phone : barcode.getContactInfo().getPhones()) {
                                    Map<String, Object> phoneMap = new HashMap<>();
                                    phoneMap.put("number", phone.getNumber());
                                    phoneMap.put("phoneType", phone.getType());
                                    queries.add(phoneMap);
                                }
                                barcodeMap.put("phones", queries);
                                queries = new ArrayList<>();
                                for (Barcode.Email email : barcode.getContactInfo().getEmails()) {
                                    Map<String, Object> emailMap = new HashMap<>();
                                    emailMap.put("address", email.getAddress());
                                    emailMap.put("body", email.getBody());
                                    emailMap.put("subject", email.getSubject());
                                    emailMap.put("emailType", email.getType());
                                    queries.add(emailMap);
                                }
                                barcodeMap.put("emails", queries);
                                List<String> urls = new ArrayList<>(barcode.getContactInfo().getUrls());
                                barcodeMap.put("urls", urls);
                                break;
                            case Barcode.TYPE_CALENDAR_EVENT:
                                barcodeMap.put("description", barcode.getCalendarEvent().getDescription());
                                barcodeMap.put("location", barcode.getCalendarEvent().getLocation());
                                barcodeMap.put("status", barcode.getCalendarEvent().getStatus());
                                barcodeMap.put("summary", barcode.getCalendarEvent().getSummary());
                                barcodeMap.put("organizer", barcode.getCalendarEvent().getOrganizer());
                                barcodeMap.put("start", barcode.getCalendarEvent().getStart().getRawValue());
                                barcodeMap.put("end", barcode.getCalendarEvent().getEnd().getRawValue());
                                break;
                        }
                        barcodeList.add(barcodeMap);
                    }
                    result.success(barcodeList);
                })
                .addOnFailureListener(e -> result.error("BarcodeDetectorError", e.toString(), null));
    }

    private void addPoints(Point[] cornerPoints, List<Map<String, Integer>> points) {
        for (Point point : cornerPoints) {
            Map<String, Integer> p = new HashMap<>();
            p.put("x", point.x);
            p.put("y", point.y);
            points.add(p);
        }
    }

    private Map<String, Integer> getBoundingPoints(@Nullable Rect rect) {
        Map<String, Integer> frame = new HashMap<>();
        if (rect == null) return frame;
        frame.put("left", rect.left);
        frame.put("right", rect.right);
        frame.put("top", rect.top);
        frame.put("bottom", rect.bottom);
        return frame;
    }

    private void closeDetector(MethodCall call) {
        String id = call.argument("id");
        com.google.mlkit.vision.barcode.BarcodeScanner barcodeScanner = instances.get(id);
        if (barcodeScanner == null) return;
        barcodeScanner.close();
        instances.remove(id);
    }
}

File: packages/google_mlkit_barcode_scanning/android/src/main/kotlin/com/google_mlkit_barcode_scanning/BarcodeScanner.kt
package com.google_mlkit_barcode_scanning

import android.content.Context
import android.graphics.Point
import android.graphics.Rect
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google_mlkit_commons.InputImageConverter
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class BarcodeScanner(
    private val context: Context,
) : MethodChannel.MethodCallHandler {
    private val instances = mutableMapOf<String, com.google.mlkit.vision.barcode.BarcodeScanner>()

    companion object {
        private const val START = "vision#startBarcodeScanner"
        private const val CLOSE = "vision#closeBarcodeScanner"
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            START -> {
                handleDetection(call, result)
            }

            CLOSE -> {
                closeDetector(call)
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initialize(call: MethodCall): com.google.mlkit.vision.barcode.BarcodeScanner {
        val formatList = call.argument<List<Int>>("formats")!!
        val options =
            if (formatList.size > 1) {
                val rest = formatList.drop(1).toIntArray()
                BarcodeScannerOptions
                    .Builder()
                    .setBarcodeFormats(formatList[0], *rest)
                    .build()
            } else {
                BarcodeScannerOptions
                    .Builder()
                    .setBarcodeFormats(formatList[0])
                    .build()
            }

        return BarcodeScanning.getClient(options)
    }

    private fun handleDetection(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val imageData = call.argument<Map<String, Any>>("imageData") ?: run {
            result.error("BarcodeDetectorError", "imageData is null", null) 
            return
        }
        
        val inputImage = InputImageConverter.getInputImageFromData(imageData, context, result) ?: return

        val id = call.argument<String>("id")!!
        val scanner = instances.getOrPut(id) { initialize(call) }

        scanner
            .process(inputImage)
            .addOnSuccessListener { barcodes ->
                val barcodeList =
                    barcodes.map { barcode ->
                        buildMap<String, Any?> {
                            val valueType = barcode.valueType
                            put("type", valueType)
                            put("format", barcode.format)
                            put("rawValue", barcode.rawValue)
                            put("rawBytes", barcode.rawBytes)
                            put("displayValue", barcode.displayValue)
                            put("rect", getBoundingPoints(barcode.boundingBox))
                            put("points", barcode.cornerPoints?.let { getPoints(it) } ?: emptyList<Any>())

                            when (valueType) {
                                Barcode.TYPE_UNKNOWN,
                                Barcode.TYPE_ISBN,
                                Barcode.TYPE_PRODUCT,
                                Barcode.TYPE_TEXT,
                                -> {
                                    Unit
                                }

                                Barcode.TYPE_WIFI -> {
                                    barcode.wifi?.let {
                                        put("ssid", it.ssid)
                                        put("password", it.password)
                                        put("encryption", it.encryptionType)
                                    }
                                }

                                Barcode.TYPE_URL -> {
                                    barcode.url?.let {
                                        put("title", it.title)
                                        put("url", it.url)
                                    }
                                }

                                Barcode.TYPE_EMAIL -> {
                                    barcode.email?.let {
                                        put("address", it.address)
                                        put("body", it.body)
                                        put("subject", it.subject)
                                        put("emailType", it.type)
                                    }
                                }

                                Barcode.TYPE_PHONE -> {
                                    barcode.phone?.let {
                                        put("number", it.number)
                                        put("phoneType", it.type)
                                    }
                                }

                                Barcode.TYPE_SMS -> {
                                    barcode.sms?.let {
                                        put("message", it.message)
                                        put("number", it.phoneNumber)
                                    }
                                }

                                Barcode.TYPE_GEO -> {
                                    barcode.geoPoint?.let {
                                        put("latitude", it.lat)
                                        put("longitude", it.lng)
                                    }
                                }

                                Barcode.TYPE_DRIVER_LICENSE -> {
                                    barcode.driverLicense?.let {
                                        put("addressCity", it.addressCity)
                                        put("addressState", it.addressState)
                                        put("addressZip", it.addressZip)
                                        put("addressStreet", it.addressStreet)
                                        put("issueDate", it.issueDate)
                                        put("birthDate", it.birthDate)
                                        put("expiryDate", it.expiryDate)
                                        put("gender", it.gender)
                                        put("licenseNumber", it.licenseNumber)
                                        put("firstName", it.firstName)
                                        put("lastName", it.lastName)
                                        put("country", it.issuingCountry)
                                    }
                                }

                                Barcode.TYPE_CONTACT_INFO -> {
                                    barcode.contactInfo?.let { contact ->
                                        put("firstName", contact.name?.first)
                                        put("lastName", contact.name?.last)
                                        put("formattedName", contact.name?.formattedName)
                                        put("organization", contact.organization)
                                        put(
                                            "addresses",
                                            contact.addresses.map { address ->
                                                mapOf(
                                                    "addressType" to address.type,
                                                    "addressLines" to address.addressLines.toList(),
                                                )
                                            },
                                        )
                                        put(
                                            "phones",
                                            contact.phones.map { phone ->
                                                mapOf("number" to phone.number, "phoneType" to phone.type)
                                            },
                                        )
                                        put(
                                            "emails",
                                            contact.emails.map { email ->
                                                mapOf(
                                                    "address" to email.address,
                                                    "body" to email.body,
                                                    "subject" to email.subject,
                                                    "emailType" to email.type,
                                                )
                                            },
                                        )
                                        put("urls", contact.urls.toList())
                                    }
                                }

                                Barcode.TYPE_CALENDAR_EVENT -> {
                                    barcode.calendarEvent?.let {
                                        put("description", it.description)
                                        put("location", it.location)
                                        put("status", it.status)
                                        put("summary", it.summary)
                                        put("organizer", it.organizer)
                                        put("start", it.start?.rawValue)
                                        put("end", it.end?.rawValue)
                                    }
                                }
                            }
                        }
                    }
                result.success(barcodeList)
            }.addOnFailureListener { e ->
                result.error("BarcodeDetectorError", e.toString(), null)
            }
    }

    private fun getPoints(cornerPoints: Array<Point>) =
        cornerPoints.map {
            mapOf("x" to it.x, "y" to it.y)
        }

    private fun getBoundingPoints(rect: Rect?) =
        rect?.let {
            mapOf("left" to it.left, "right" to it.right, "top" to it.top, "bottom" to it.bottom)
        } ?: emptyMap<String, Int>()

    private fun closeDetector(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        instances.remove(id)?.close()
    }
}
